{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.11.3\n"
     ]
    }
   ],
   "source": [
    "import anndata as ad\n",
    "import numpy as np \n",
    "import pandas as pd\n",
    "import scipy\n",
    "from scipy.sparse import csr_matrix\n",
    "import scanpy as sc\n",
    "print(ad.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in data\n",
    "data = ad.read_h5ad(r'/root/capsule/data/AD-Group/MTG/balanced_SEAAD_MTG_RNAseq_Singleome_final-nuclei.2024-06-18.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 139000 × 36601\n",
      "    obs: 'sample_id', 'Neurotypical reference', 'Donor ID', 'Organism', 'Brain Region', 'Sex', 'Gender', 'Age at Death', 'Race (choice=White)', 'Race (choice=Black/ African American)', 'Race (choice=Asian)', 'Race (choice=American Indian/ Alaska Native)', 'Race (choice=Native Hawaiian or Pacific Islander)', 'Race (choice=Unknown or unreported)', 'Race (choice=Other)', 'specify other race', 'Hispanic/Latino', 'Highest level of education', 'Years of education', 'PMI', 'Fresh Brain Weight', 'Brain pH', 'Overall AD neuropathological Change', 'Thal', 'Braak', 'CERAD score', 'Overall CAA Score', 'Highest Lewy Body Disease', 'Total Microinfarcts (not observed grossly)', 'Total microinfarcts in screening sections', 'Atherosclerosis', 'Arteriolosclerosis', 'LATE', 'Cognitive Status', 'Last CASI Score', 'Interval from last CASI in months', 'Last MMSE Score', 'Interval from last MMSE in months', 'Last MOCA Score', 'Interval from last MOCA in months', 'APOE Genotype', 'Primary Study Name', 'Secondary Study Name', 'NeuN positive fraction on FANS', 'RIN', 'cell_prep_type', 'facs_population_plan', 'rna_amplification', 'sample_name', 'sample_quantity_count', 'expc_cell_capture', 'method', 'pcr_cycles', 'percent_cdna_longer_than_400bp', 'rna_amplification_pass_fail', 'amplified_quantity_ng', 'load_name', 'library_prep', 'library_input_ng', 'r1_index', 'avg_size_bp', 'quantification_fmol', 'library_prep_pass_fail', 'exp_component_vendor_name', 'batch_vendor_name', 'experiment_component_failed', 'alignment', 'Genome', 'ar_id', 'bc', 'GEX_Estimated_number_of_cells', 'GEX_number_of_reads', 'GEX_sequencing_saturation', 'GEX_Mean_raw_reads_per_cell', 'GEX_Q30_bases_in_barcode', 'GEX_Q30_bases_in_read_2', 'GEX_Q30_bases_in_UMI', 'GEX_Percent_duplicates', 'GEX_Q30_bases_in_sample_index_i1', 'GEX_Q30_bases_in_sample_index_i2', 'GEX_Reads_with_TSO', 'GEX_Sequenced_read_pairs', 'GEX_Valid_UMIs', 'GEX_Valid_barcodes', 'GEX_Reads_mapped_to_genome', 'GEX_Reads_mapped_confidently_to_genome', 'GEX_Reads_mapped_confidently_to_intergenic_regions', 'GEX_Reads_mapped_confidently_to_intronic_regions', 'GEX_Reads_mapped_confidently_to_exonic_regions', 'GEX_Reads_mapped_confidently_to_transcriptome', 'GEX_Reads_mapped_antisense_to_gene', 'GEX_Fraction_of_transcriptomic_reads_in_cells', 'GEX_Total_genes_detected', 'GEX_Median_UMI_counts_per_cell', 'GEX_Median_genes_per_cell', 'Multiome_Feature_linkages_detected', 'Multiome_Linked_genes', 'Multiome_Linked_peaks', 'ATAC_Confidently_mapped_read_pairs', 'ATAC_Fraction_of_genome_in_peaks', 'ATAC_Fraction_of_high_quality_fragments_in_cells', 'ATAC_Fraction_of_high_quality_fragments_overlapping_TSS', 'ATAC_Fraction_of_high_quality_fragments_overlapping_peaks', 'ATAC_Fraction_of_transposition_events_in_peaks_in_cells', 'ATAC_Mean_raw_read_pairs_per_cell', 'ATAC_Median_high_quality_fragments_per_cell', 'ATAC_Non-nuclear_read_pairs', 'ATAC_Number_of_peaks', 'ATAC_Percent_duplicates', 'ATAC_Q30_bases_in_barcode', 'ATAC_Q30_bases_in_read_1', 'ATAC_Q30_bases_in_read_2', 'ATAC_Q30_bases_in_sample_index_i1', 'ATAC_Sequenced_read_pairs', 'ATAC_TSS_enrichment_score', 'ATAC_Unmapped_read_pairs', 'ATAC_Valid_barcodes', 'Number of mapped reads', 'Number of unmapped reads', 'Number of multimapped reads', 'Number of reads', 'Number of UMIs', 'Genes detected', 'Doublet score', 'Fraction mitochondrial UMIs', 'Used in analysis', 'Class confidence', 'Class', 'Subclass confidence', 'Subclass', 'Supertype confidence', 'Supertype (non-expanded)', 'Supertype', 'Continuous Pseudo-progression Score', 'Severely Affected Donor'\n",
      "    obsm: 'X_scVI', 'X_umap'\n",
      "    layers: 'UMIs'\n",
      "Shape: (139000, 36601)\n"
     ]
    }
   ],
   "source": [
    "print(data)\n",
    "print(f'Shape: {data.shape}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Cell_0', 'Cell_1', 'Cell_2', 'Cell_3', 'Cell_4'], dtype='object')\n",
      "Index(['MIR1302-2HG', 'FAM138A', 'OR4F5', 'AL627309.1', 'AL627309.3'], dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(data.obs_names[:5])\n",
    "print(data.var_names[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                             sample_id Neurotypical reference  \\\n",
      "Cell_0  CATCAAGTCGGATACT-L8TX_210729_01_G12-1153814338                  False   \n",
      "Cell_1  TGCTGAAAGCGATCGA-L8TX_210415_01_E11-1153814186                  False   \n",
      "Cell_2             AAGCCATGTGATATAG-LKTX_191204_01_F01                   True   \n",
      "Cell_3  TCGATTTGTCGTGATT-L8TX_210715_01_D05-1153814295                  False   \n",
      "Cell_4  CGCCAGATCAACCTTT-L8TX_210722_01_B07-1153814297                  False   \n",
      "\n",
      "          Donor ID Organism Brain Region     Sex  Gender  Age at Death  \\\n",
      "Cell_0  H21.33.035    human    Human MTG  Female  Female          97.0   \n",
      "Cell_1  H21.33.007    human    Human MTG  Female  Female          86.0   \n",
      "Cell_2  H19.30.001    human    Human MTG    Male    Male           NaN   \n",
      "Cell_3  H20.33.032    human    Human MTG    Male    Male          98.0   \n",
      "Cell_4  H20.33.034    human    Human MTG  Female  Female          85.0   \n",
      "\n",
      "       Race (choice=White) Race (choice=Black/ African American)  ...  \\\n",
      "Cell_0             Checked                             Unchecked  ...   \n",
      "Cell_1             Checked                             Unchecked  ...   \n",
      "Cell_2           Reference                             Reference  ...   \n",
      "Cell_3             Checked                             Unchecked  ...   \n",
      "Cell_4             Checked                             Unchecked  ...   \n",
      "\n",
      "       Used in analysis Class confidence                        Class  \\\n",
      "Cell_0             True              1.0  Non-neuronal and Non-neural   \n",
      "Cell_1             True              1.0  Non-neuronal and Non-neural   \n",
      "Cell_2             True              1.0  Non-neuronal and Non-neural   \n",
      "Cell_3             True              1.0  Non-neuronal and Non-neural   \n",
      "Cell_4             True              1.0  Non-neuronal and Non-neural   \n",
      "\n",
      "       Subclass confidence         Subclass Supertype confidence  \\\n",
      "Cell_0            1.000000  Oligodendrocyte             1.000000   \n",
      "Cell_1            1.000000  Oligodendrocyte             1.000000   \n",
      "Cell_2            1.000000  Oligodendrocyte             1.000000   \n",
      "Cell_3            0.999693  Oligodendrocyte             0.999982   \n",
      "Cell_4            1.000000  Oligodendrocyte             1.000000   \n",
      "\n",
      "       Supertype (non-expanded) Supertype Continuous Pseudo-progression Score  \\\n",
      "Cell_0                  Oligo_4   Oligo_4                            0.376970   \n",
      "Cell_1                  Oligo_4   Oligo_4                            0.723364   \n",
      "Cell_2                  Oligo_4   Oligo_4                                 NaN   \n",
      "Cell_3                  Oligo_4   Oligo_4                            0.818041   \n",
      "Cell_4                  Oligo_4   Oligo_4                            0.368330   \n",
      "\n",
      "       Severely Affected Donor  \n",
      "Cell_0                       N  \n",
      "Cell_1                       N  \n",
      "Cell_2                     NaN  \n",
      "Cell_3                       N  \n",
      "Cell_4                       N  \n",
      "\n",
      "[5 rows x 135 columns]\n",
      "Index(['sample_id', 'Neurotypical reference', 'Donor ID', 'Organism',\n",
      "       'Brain Region', 'Sex', 'Gender', 'Age at Death', 'Race (choice=White)',\n",
      "       'Race (choice=Black/ African American)',\n",
      "       ...\n",
      "       'Used in analysis', 'Class confidence', 'Class', 'Subclass confidence',\n",
      "       'Subclass', 'Supertype confidence', 'Supertype (non-expanded)',\n",
      "       'Supertype', 'Continuous Pseudo-progression Score',\n",
      "       'Severely Affected Donor'],\n",
      "      dtype='object', length=135)\n"
     ]
    }
   ],
   "source": [
    "# obs\n",
    "print(data.obs.head())\n",
    "print(data.obs.columns)\n",
    "\n",
    "df = pd.DataFrame(data = data.obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "ename": "MemoryError",
     "evalue": "Unable to allocate 19.0 GiB for an array with shape (139000, 36601) and data type float32",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mMemoryError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[72], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(\u001b[43mdata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mX\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtoarray\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mif\u001b[39;00m scipy\u001b[38;5;241m.\u001b[39msparse\u001b[38;5;241m.\u001b[39missparse(data\u001b[38;5;241m.\u001b[39mX) \u001b[38;5;28;01melse\u001b[39;00m data\u001b[38;5;241m.\u001b[39mX, \n\u001b[1;32m      2\u001b[0m                  index\u001b[38;5;241m=\u001b[39mdata\u001b[38;5;241m.\u001b[39mobs_names, \n\u001b[1;32m      3\u001b[0m                  columns\u001b[38;5;241m=\u001b[39mdata\u001b[38;5;241m.\u001b[39mvar_names)\n\u001b[1;32m      5\u001b[0m df\u001b[38;5;241m.\u001b[39mhead()\n",
      "File \u001b[0;32m/opt/conda/lib/python3.10/site-packages/scipy/sparse/_compressed.py:1170\u001b[0m, in \u001b[0;36m_cs_matrix.toarray\u001b[0;34m(self, order, out)\u001b[0m\n\u001b[1;32m   1168\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m out \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m order \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1169\u001b[0m     order \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_swap(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcf\u001b[39m\u001b[38;5;124m'\u001b[39m)[\u001b[38;5;241m0\u001b[39m]\n\u001b[0;32m-> 1170\u001b[0m out \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_process_toarray_args\u001b[49m\u001b[43m(\u001b[49m\u001b[43morder\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1171\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (out\u001b[38;5;241m.\u001b[39mflags\u001b[38;5;241m.\u001b[39mc_contiguous \u001b[38;5;129;01mor\u001b[39;00m out\u001b[38;5;241m.\u001b[39mflags\u001b[38;5;241m.\u001b[39mf_contiguous):\n\u001b[1;32m   1172\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mOutput array must be C or F contiguous\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m/opt/conda/lib/python3.10/site-packages/scipy/sparse/_base.py:1367\u001b[0m, in \u001b[0;36m_spbase._process_toarray_args\u001b[0;34m(self, order, out)\u001b[0m\n\u001b[1;32m   1365\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m out\n\u001b[1;32m   1366\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1367\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mzeros\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshape\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43morder\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43morder\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mMemoryError\u001b[0m: Unable to allocate 19.0 GiB for an array with shape (139000, 36601) and data type float32"
     ]
    }
   ],
   "source": [
    "df = pd.DataFrame(data.X.toarray() if scipy.sparse.issparse(data.X) else data.X, \n",
    "                 index=data.obs_names, \n",
    "                 columns=data.var_names)\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_____"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "exp_component_name\n",
       "CATCAAGTCGGATACT-L8TX_210729_01_G12-1153814338                          Oligo_4\n",
       "TGCTGAAAGCGATCGA-L8TX_210415_01_E11-1153814186                          Oligo_4\n",
       "AAGCCATGTGATATAG-LKTX_191204_01_F01-1156636535                          Oligo_4\n",
       "TCGATTTGTCGTGATT-L8TX_210715_01_D05-1153814295                          Oligo_4\n",
       "CGCCAGATCAACCTTT-L8TX_210722_01_B07-1153814297                          Oligo_4\n",
       "                                                                         ...   \n",
       "TTCATGTTCAGTCACA-L8TX_210430_01_C04-1142430413_aug_2cells_comb_814-1     Endo_1\n",
       "ATCGCCTAGTCTGCAT-L8TX_210430_01_D05-1142430407_aug_2cells_comb_815-1     Endo_1\n",
       "CCGAACGAGAGTCTGG-L8TX_210430_01_C04-1142430413_aug_2cells_comb_816-1     Endo_1\n",
       "GTGAGGATCACTCCGT-L8TX_201030_01_B12-1142430231_aug_2cells_comb_817-1     Endo_1\n",
       "CGCCAGAAGGTACCTT-L8TX_210729_01_C12-1153814340_aug_2cells_comb_818-1     Endo_1\n",
       "Name: Supertype, Length: 139000, dtype: category\n",
       "Categories (139, object): ['Astro_1', 'Astro_2', 'Astro_3', 'Astro_4', ..., 'Vip_18', 'Vip_19', 'Vip_21', 'Vip_23']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.obs['Supertype']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Cell_0', 'Cell_1', 'Cell_2', 'Cell_3', 'Cell_4', 'Cell_5', 'Cell_6',\n",
      "       'Cell_7', 'Cell_8', 'Cell_9'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "data.obs_names = [f\"Cell_{i:d}\" for i in range(data.n_obs)]\n",
    "data.var_names = [f\"Gene_{i:d}\" for i in range(data.n_vars)]\n",
    "print(data.obs_names[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_0</th>\n",
       "      <th>Gene_1</th>\n",
       "      <th>Gene_2</th>\n",
       "      <th>Gene_3</th>\n",
       "      <th>Gene_4</th>\n",
       "      <th>Gene_5</th>\n",
       "      <th>Gene_6</th>\n",
       "      <th>Gene_7</th>\n",
       "      <th>Gene_8</th>\n",
       "      <th>Gene_9</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene_36591</th>\n",
       "      <th>Gene_36592</th>\n",
       "      <th>Gene_36593</th>\n",
       "      <th>Gene_36594</th>\n",
       "      <th>Gene_36595</th>\n",
       "      <th>Gene_36596</th>\n",
       "      <th>Gene_36597</th>\n",
       "      <th>Gene_36598</th>\n",
       "      <th>Gene_36599</th>\n",
       "      <th>Gene_36600</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cell_0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_138995</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_138996</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_138997</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_138998</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell_138999</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>139000 rows × 36601 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Gene_0  Gene_1  Gene_2  Gene_3  Gene_4  Gene_5  Gene_6  Gene_7  \\\n",
       "Cell_0          0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Cell_1          0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Cell_2          0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Cell_3          0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Cell_4          0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "...             ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "Cell_138995     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Cell_138996     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Cell_138997     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Cell_138998     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "Cell_138999     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "             Gene_8  Gene_9  ...  Gene_36591  Gene_36592  Gene_36593  \\\n",
       "Cell_0          0.0     0.0  ...         0.0         0.0         0.0   \n",
       "Cell_1          0.0     0.0  ...         0.0         0.0         0.0   \n",
       "Cell_2          0.0     0.0  ...         0.0         0.0         0.0   \n",
       "Cell_3          0.0     0.0  ...         0.0         0.0         0.0   \n",
       "Cell_4          0.0     0.0  ...         0.0         0.0         0.0   \n",
       "...             ...     ...  ...         ...         ...         ...   \n",
       "Cell_138995     0.0     0.0  ...         0.0         0.0         0.0   \n",
       "Cell_138996     0.0     0.0  ...         0.0         0.0         0.0   \n",
       "Cell_138997     0.0     0.0  ...         0.0         0.0         0.0   \n",
       "Cell_138998     0.0     0.0  ...         0.0         0.0         0.0   \n",
       "Cell_138999     0.0     0.0  ...         0.0         0.0         0.0   \n",
       "\n",
       "             Gene_36594  Gene_36595  Gene_36596  Gene_36597  Gene_36598  \\\n",
       "Cell_0              0.0         0.0         0.0         0.0         0.0   \n",
       "Cell_1              0.0         0.0         0.0         0.0         0.0   \n",
       "Cell_2              0.0         0.0         0.0         0.0         0.0   \n",
       "Cell_3              0.0         0.0         0.0         0.0         0.0   \n",
       "Cell_4              0.0         0.0         0.0         0.0         0.0   \n",
       "...                 ...         ...         ...         ...         ...   \n",
       "Cell_138995         0.0         0.0         0.0         0.0         0.0   \n",
       "Cell_138996         0.0         0.0         0.0         0.0         0.0   \n",
       "Cell_138997         0.0         0.0         0.0         0.0         0.0   \n",
       "Cell_138998         0.0         0.0         0.0         0.0         0.0   \n",
       "Cell_138999         0.0         0.0         0.0         0.0         0.0   \n",
       "\n",
       "             Gene_36599  Gene_36600  \n",
       "Cell_0              0.0         0.0  \n",
       "Cell_1              0.0         0.0  \n",
       "Cell_2              0.0         0.0  \n",
       "Cell_3              0.0         0.0  \n",
       "Cell_4              0.0         0.0  \n",
       "...                 ...         ...  \n",
       "Cell_138995         0.0         0.0  \n",
       "Cell_138996         0.0         0.0  \n",
       "Cell_138997         0.0         0.0  \n",
       "Cell_138998         0.0         0.0  \n",
       "Cell_138999         0.0         0.0  \n",
       "\n",
       "[139000 rows x 36601 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.to_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
